cmd/carbonapi/http/find_handlers.go

```diff
@@ -43,7 +44,71 @@ type treejson struct {
 
 var treejsonContext = make(map[string]int)
 
-var queryStore = make(map[string]interface{})
+var queryStore = NewCacheQueries()
+
+type CacheQueries struct {
+	mu      sync.RWMutex
+	queries map[string]CacheElement
+}
+
+func NewCacheElement(signature string, deph int32, request pbv3.MultiGlobRequest) CacheElement {
+	return CacheElement{
+		QueriesSignature: signature,
+		Request:          request,
+		Deph:             deph,
+		LastAccess:       time.Now(),
+	}
+}
+
+func NewCacheQueries() CacheQueries {
+	return CacheQueries{
+		mu:      sync.RWMutex{},
+		queries: make(map[string]CacheElement),
+	}
+}
+
+type CacheElement struct {
+	QueriesSignature string
+	Request          pbv3.MultiGlobRequest
+	Deph             int32
+	LastAccess       time.Time
+}
+
+func (c *CacheQueries) Set(el CacheElement) {
+	c.mu.Lock()
+	c.queries[el.QueriesSignature] = el
+	c.mu.Unlock()
+}
+
+func (c *CacheQueries) Get(key string) (CacheElement, bool) {
+	c.mu.RLock()
+	val, ok := c.queries[key]
+	c.mu.RUnlock()
+	return val, ok
+}
+
+func (c *CacheQueries) GetUpdate(key string) (CacheElement, bool) {
+	c.mu.RLock()
+	val, ok := c.queries[key]
+	c.mu.RUnlock()
+	c.mu.Lock()
+	if ok {
```

Comment:
need to recheck if the val exist as the map may have been updated between runlock & lock